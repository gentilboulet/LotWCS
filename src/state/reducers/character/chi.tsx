import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/chi";
import { TChiState } from "../../models/character/chi";

import { increase } from "../../models/character/chi";

export type IChiAction = ActionType<typeof actions>;

export const chiReducer = produce(
  (draft: Draft<TChiState>, action: IChiAction) => {
    switch (action.type) {
      case getType(actions.chiBuy):
        increase(draft, action.payload.chi, action.payload.value);
        break;
      default:
    }
  },
);
