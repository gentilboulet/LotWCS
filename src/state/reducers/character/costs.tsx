import { produce, Draft } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/costs";
import { applyCost } from "../../models/character/costs";
import { ICharacterState } from "../../models/character";

export type ICostAction = ActionType<typeof actions>;

export const costReducer = produce(
  (draft: Draft<ICharacterState>, action: ICostAction) => {
    switch (action.type) {
      case getType(actions.payCost):
        applyCost(draft, action.payload.cost);
        break;
      default:
    }
  },
);
