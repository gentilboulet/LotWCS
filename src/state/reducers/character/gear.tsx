import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/gear";
import { TGearState } from "../../models/character/gear";

export const gearReducer = produce(
  (draft: Draft<TGearState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.armorBuy):
        draft.armors.push({
          description: action.payload.description,
          name: action.payload.name,
          type: action.payload.type,
        });
        break;
      case getType(actions.armorCustomize):
        draft.armors[action.payload.armorIdx].description =
          action.payload.description;
        break;
      case getType(actions.weaponBuy):
        draft.weapons.push({
          description: action.payload.description,
          name: action.payload.name,
          type: action.payload.type,
        });
        break;
      case getType(actions.weaponCustomize):
        draft.weapons[action.payload.weaponIdx].description =
          action.payload.description;
        break;
      default:
    }
  },
);
