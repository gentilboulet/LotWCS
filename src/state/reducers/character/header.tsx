import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/header";
import { ICharacterState } from "../../models/character";
import { setRank } from "../../models/character/header";

export const headerReducer = produce(
  (draft: Draft<ICharacterState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.setName):
        draft.name = action.payload.name;
        break;
      case getType(actions.setConcept):
        draft.concept = action.payload.concept;
        break;
      case getType(actions.setRank):
        setRank(draft, action.payload.rank);
        break;
      case getType(actions.setArchetype):
        draft.archetype = action.payload.archetype;
        break;
      default:
    }
  },
);
