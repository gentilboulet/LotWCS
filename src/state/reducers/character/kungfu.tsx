import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/kungfu";
import {
  addKungFuTechnique,
  IKungFuState,
  openStyle,
} from "../../models/character/kungfu";

export const kungfuReducer = produce(
  (draft: Draft<IKungFuState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.openStyle):
        openStyle(draft, action.payload.kungfuType, action.payload.uid);
        break;
      case getType(actions.buyTechnique):
        addKungFuTechnique(
          draft,
          action.payload.kungfuType,
          action.payload.styleUid,
          action.payload.uid,
        );
        break;
      case getType(actions.customStyleName):
      case getType(actions.customTechniqueName):
      default:
    }
  },
);
