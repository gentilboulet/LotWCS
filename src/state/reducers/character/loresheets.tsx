import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/loresheets";
import {
  buyLoresheetOption,
  ILoresheetsState,
  openLoresheet,
} from "../../models/character/loresheets";

export const loresheetsReducer = produce(
  (draft: Draft<ILoresheetsState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.open):
        openLoresheet(draft, action.payload.uid);
        break;
      case getType(actions.buyOption):
        buyLoresheetOption(
          draft,
          action.payload.lsUid,
          action.payload.uid,
          action.payload.payload,
        );
        break;
      default:
    }
  },
);
