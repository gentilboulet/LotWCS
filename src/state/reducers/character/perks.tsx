import { Draft, produce } from "immer";
import { ActionType, getType } from "typesafe-actions";

import * as actions from "../../actions/character/perks";
import { ICharacterState } from "../../models/character";
import { applyBonuses } from "../../models/character/bonuses";
import { pushToDiscounts } from "../../models/character/discounts";

export const perksReducer = produce(
  (draft: Draft<ICharacterState>, action: ActionType<typeof actions>) => {
    switch (action.type) {
      case getType(actions.applyBonus):
        applyBonuses(draft, action.payload.bonuses);
        break;
      case getType(actions.pushDiscount):
        pushToDiscounts(draft, action.payload.discounts);
        break;
      default:
    }
  },
);
