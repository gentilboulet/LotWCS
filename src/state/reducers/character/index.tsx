import { produce, Draft } from "immer";
import { isActionOf } from "typesafe-actions";

import { ICharacterAction } from "../../actions/character";
import { ICharacterState } from "../../models/character";

// Sub Reducers
import { chiReducer } from "./chi";
import { costReducer } from "./costs";
import { headerReducer } from "./header";
import { kungfuReducer } from "./kungfu";
import { loresheetsReducer } from "./loresheets";
import { perksReducer } from "./perks";
import { skillsReducer } from "./skills";
import { virtuesReducer } from "./virtues";

import * as chi from "../../actions/character/chi";
import * as costs from "../../actions/character/costs";
import * as header from "../../actions/character/header";
import * as kungfu from "../../actions/character/kungfu";
import * as loresheets from "../../actions/character/loresheets";
import * as perks from "../../actions/character/perks";
import * as skills from "../../actions/character/skills";
import * as virtues from "../../actions/character/virtues";

export const globalReducer = produce(
  (draft: Draft<ICharacterState>, action: ICharacterAction) => {
    if (isActionOf(Object.values(chi), action)) {
      draft.chi = chiReducer(draft.chi, action);
    }
    if (isActionOf(Object.values(costs), action)) {
      costReducer(draft, action);
    }
    if (isActionOf(Object.values(header), action)) {
      headerReducer(draft, action);
    }
    if (isActionOf(Object.values(kungfu), action)) {
      draft.kungfu = kungfuReducer(draft.kungfu, action);
    }
    if (isActionOf(Object.values(loresheets), action)) {
      draft.loresheets = loresheetsReducer(draft.loresheets, action);
    }
    if (isActionOf(Object.values(perks), action)) {
      perksReducer(draft, action);
    }
    if (isActionOf(Object.values(skills), action)) {
      draft.skills = skillsReducer(draft.skills, action);
    }
    if (isActionOf(Object.values(virtues), action)) {
      draft.virtues = virtuesReducer(draft.virtues, action);
    }
  },
);
