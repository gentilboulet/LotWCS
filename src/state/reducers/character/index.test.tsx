import { ICharacterAction } from "../../actions/character";
import { initialStateFactory } from "../../models/character";
import { globalReducer } from "../../reducers/character";

const initialState = initialStateFactory();

describe("Testing globalReducer", () => {
  it("should create an initial state when used with no state", () => {
    expect(initialState).toMatchSnapshot();
    const junk = { type: "JUNK_ACTION" };
    expect(globalReducer(initialState, junk as ICharacterAction)).toMatchObject(
      initialState,
    );

    const action = ({
      payload: { name },
      type: "header/SET_NAME",
    } as unknown) as ICharacterAction;
    const expectedState = globalReducer(initialState, action);
    expect(globalReducer(initialState, action)).toMatchObject(expectedState);
  });

  it("should do nothing with a junk action", () => {
    expect(initialState).toMatchSnapshot();
    const junk = { type: "JUNK_ACTION" };
    expect(globalReducer(initialState, junk as ICharacterAction)).toMatchObject(
      initialState,
    );
  });
});
