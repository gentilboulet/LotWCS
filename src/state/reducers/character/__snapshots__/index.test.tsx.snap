// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing globalReducer should create an initial state when used with no state 1`] = `
Object {
  "archetype": undefined,
  "automatics": Array [],
  "chi": Object {
    "corrupt": Object {
      "cultivation": 0,
      "value": 0,
    },
    "earth": Object {
      "cultivation": 0,
      "value": 0,
    },
    "enlightened": Object {
      "cultivation": 0,
      "value": 0,
    },
    "fire": Object {
      "cultivation": 0,
      "value": 0,
    },
    "general": Object {
      "cultivation": 0,
      "value": 0,
    },
    "metal": Object {
      "cultivation": 0,
      "value": 0,
    },
    "water": Object {
      "cultivation": 0,
      "value": 0,
    },
    "wood": Object {
      "cultivation": 0,
      "value": 0,
    },
  },
  "concept": undefined,
  "destiny": 0,
  "discounts": Array [],
  "entanglement": 0,
  "gear": Object {
    "armors": Array [],
    "weapons": Array [],
  },
  "kungfu": Object {
    "KUNGFU_EXTERNAL": Object {},
    "KUNGFU_INTERNAL": Object {},
    "noRestrictionInternal": Array [],
  },
  "loresheets": Object {},
  "name": undefined,
  "rank": undefined,
  "skills": Array [
    Object {
      "name": "Awareness",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Confidence",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Crafting",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Finesse",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Hardiness",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Inspire",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Learning",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Medicine",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Might",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Perform",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Politics",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Ride",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Stealth",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Survival",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Tactics",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Wu Wei",
      "specialities": Array [],
      "value": 0,
    },
  ],
  "virtues": Array [
    Object {
      "name": "Honor",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Benevolence",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Righteousness",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Loyalty",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Force",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Revenge",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Individualism",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Obsession",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ruthlessness",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ferocity",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
  ],
}
`;

exports[`Testing globalReducer should do nothing with a junk action 1`] = `
Object {
  "archetype": undefined,
  "automatics": Array [],
  "chi": Object {
    "corrupt": Object {
      "cultivation": 0,
      "value": 0,
    },
    "earth": Object {
      "cultivation": 0,
      "value": 0,
    },
    "enlightened": Object {
      "cultivation": 0,
      "value": 0,
    },
    "fire": Object {
      "cultivation": 0,
      "value": 0,
    },
    "general": Object {
      "cultivation": 0,
      "value": 0,
    },
    "metal": Object {
      "cultivation": 0,
      "value": 0,
    },
    "water": Object {
      "cultivation": 0,
      "value": 0,
    },
    "wood": Object {
      "cultivation": 0,
      "value": 0,
    },
  },
  "concept": undefined,
  "destiny": 0,
  "discounts": Array [],
  "entanglement": 0,
  "gear": Object {
    "armors": Array [],
    "weapons": Array [],
  },
  "kungfu": Object {
    "KUNGFU_EXTERNAL": Object {},
    "KUNGFU_INTERNAL": Object {},
    "noRestrictionInternal": Array [],
  },
  "loresheets": Object {},
  "name": undefined,
  "rank": undefined,
  "skills": Array [
    Object {
      "name": "Awareness",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Confidence",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Crafting",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Finesse",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Hardiness",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Inspire",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Learning",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Medicine",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Might",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Perform",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Politics",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Ride",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Stealth",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Survival",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Tactics",
      "specialities": Array [],
      "value": 0,
    },
    Object {
      "name": "Wu Wei",
      "specialities": Array [],
      "value": 0,
    },
  ],
  "virtues": Array [
    Object {
      "name": "Honor",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Benevolence",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Righteousness",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Loyalty",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Force",
      "type": "VIRTUE_CHIVALROUS",
      "value": 0,
    },
    Object {
      "name": "Revenge",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Individualism",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Obsession",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ruthlessness",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
    Object {
      "name": "Ferocity",
      "type": "VIRTUE_SELFISH",
      "value": 0,
    },
  ],
}
`;
